middleware/proxy/grpc.go

```diff
@@ -0,0 +1,70 @@
+package proxy
+
+import (
+	"crypto/tls"
+	"log"
+
+	"github.com/miekg/coredns/middleware/proxy/pb"
+	"github.com/miekg/coredns/request"
+
+	"github.com/miekg/dns"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"golang.org/x/net/context"
+
+)
+
+type grpcClient struct {
+	dialOpt grpc.DialOption
+	clients	map[string]pb.DnsServiceClient
+	upstream *staticUpstream
+}
+
+func newGrpcClient(tls *tls.Config, u *staticUpstream) *grpcClient {
+	g := &grpcClient{upstream: u}
+
+	if tls == nil {
+		g.dialOpt = grpc.WithInsecure()
+	} else {
+		g.dialOpt = grpc.WithTransportCredentials(credentials.NewTLS(tls))
+	}
+	g.clients = map[string]pb.DnsServiceClient{}
+
+	return g
+}
+
+func (g *grpcClient) Exchange(addr string, state request.Request) (*dns.Msg, error) {
+	msg, err := state.Req.Pack()
+	if err != nil {
+		return nil, err
+	}
+
+	reply, err := g.clients[addr].Query(context.TODO(), &pb.DnsPacket{Msg: msg})
+	if err != nil {
+		return nil, err
+	}
+	d := new(dns.Msg)
+	err = d.Unpack(reply.Msg)
+	if err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+func (g *grpcClient) Protocol() string { return "grpc" }
+
+func (g *grpcClient) OnShutdown(p *Proxy) error {
+	return nil
+}
+
+func (g *grpcClient) OnStartup(p *Proxy) error {
+	for _, host := range g.upstream.Hosts {
+		conn, err := grpc.Dial(host.Name, g.dialOpt)
+		if err != nil {
+			log.Printf("[WARNING] Skipping gRPC host '%s' due to Dial error: '%s'", host, err)
+		} else {
+			g.clients[host.Name] = pb.NewDnsServiceClient(conn)
```

Comment:
Ok, so here is the issue. The OnShutdown is not being called when I do a reload. This is because the port bind is failing, so the Restart method is bailing partway through:

```
2017/02/10 15:55:00 [INFO] Reloading
2017/02/10 15:55:00 [INFO] Creating gRPC connection 0 to 127.0.0.1:4443.
2017/02/10 15:55:00 [INFO] Creating gRPC connection 1 to 127.0.0.1:4444.
2017/02/10 15:55:00 [ERROR] SIGUSR1: listen tcp :1053: bind: address already in use
```

It seems the issue stems from this statement here: https://github.com/mholt/caddy/blob/master/caddy.go#L619 - it seems dnsserver.Server is NOT implementing the GracefulServer interface: https://godoc.org/github.com/mholt/caddy#GracefulServer